client/src/main/java/org/eclipse/hono/client/impl/HonoConnectionImpl.java

```diff
@@ -436,8 +441,14 @@ private void connect(
                                             connectionFactory.getPort(),
                                             connectionFactory.getServerRole(),
                                             newConnection.getRemoteContainer());
-                                    setConnection(newConnection);
-                                    wrappedConnectionHandler.handle(Future.succeededFuture(this));
+                                    createDefaultSession(newConnection)
+                                        .onSuccess(newSession -> {
+                                            setConnection(newConnection, newSession);
+                                            wrappedConnectionHandler.handle(Future.succeededFuture(this));
+                                        })
+                                        .onFailure(t -> {
+                                            wrappedConnectionHandler.handle(Future.failedFuture(t));
```

Comment:
Instead of failing all reconnect attempts like this, close the connection and call `reconnect` in the `onFailure` case here instead?

Reply:
If the peer does not accept the session I do not really see why it should accept it in subsequent attempts. FMPOV this is comparable to wrong credentials, the peer either is willing to open a session or not, or am I mistaken?
I agree that we should close the newly opened connection but then I guess we will need to fail the connection attempt. WDYT?

Reply:
For a session to not get established by the remote peer, here Qpid in particular, I could think of 2 reasons:
- maxSessions having been reached already (shouldn't happen here since we only use one session per connection)
- or Qpid not having enough memory to create a new session.
For the 2nd case, I think it would be good to retry the connection attempt, potentially ending up at another Qpid instance or Qpid having been restarted then.
Failing the connection attempt instead in this case would be fatal and cause the whole adapter to need a restart in case of the telemetry/event sender connection for example, if I'm not mistaken.

Reply:
I see your point and it makes sense to me. The default session in vertx-proton only has a close handler which is used to trigger the underlying connection's close handler. This mechanism is already being used by Hono implicitly, so I guess we could use the same approach here as well, i.e. do not register an explicit open handler for the session at all but simply rely on the (standard) behavior of the connection's close handler to trigger reconnect. WDYT?

Reply:
Yes, I agree.